horovod/run/elastic/driver.py

```diff
@@ -174,14 +172,14 @@ def _notify_workers_host_changes(self):
                           .format(host, slot))
 
     def _update_host_assignments(self):
-        # Update the list of hosts that are currently assigned workers.
+        # Determine the slots that are already filled so we do not respawn these processes
+        active_slots = set([(host, slot_info.local_rank) for host, slot_info in self._host_assignments.items()])
+
         # We need to ensure this list preserves relative order to ensure the oldest hosts are assigned lower ranks.
-        new_assigned_hosts = []
         self._assigned_hosts = self._discovered_hosts.filter_available_hosts(self._assigned_hosts)
         current_hosts = set(self._assigned_hosts)
         for host in self.get_available_hosts():
             if host not in current_hosts:
-                new_assigned_hosts.append(host)
                 self._assigned_hosts.append(host)
 
         # Adjust the host assignments to account for added / removed hosts
```

Comment:
When old hosts get additional slots then some hosts later in `self._assigned_hosts` will drop out of `self._host_assignments`, but their processes will continue running. I think I saw some code that handles this situation where processes do not get a rank though they are running. So they are turned idle I guess.

Reply:
I think `self._assigned_hosts` should be renamed as the naming is too similar to `self._host_assignment`. I think `self._available_hosts` describes it best (as they are populated via `self._assigned_hosts = self._discovered_hosts.filter_available_hosts(self._assigned_hosts)` plus `self.get_available_hosts()`).

Reply:
You mentioned that code in one of your comments:

> We do already notify workers if an assigned host is removed from the available hosts, at which point it can be gracefully removed without needing to fail on allreduce. The worker will kill itself after rendezvous when it received -1 for its rank.

So this is handled.

Reply:
Renamed to `ordered_available_hosts` for clarity.  Let me know if that makes sense.

Reply:
very good name!